velox/connectors/hive/HiveDataSink.cpp

```diff
@@ -70,22 +244,151 @@ std::unique_ptr<velox::dwrf::Writer> HiveDataSink::createWriter() {
   // Without explicitly setting flush policy, the default memory based flush
   // policy is used.
 
-  auto hiveWriterParameters =
-      std::dynamic_pointer_cast<const HiveWriterParameters>(
-          writeProtocol_->getWriterParameters(
-              insertTableHandle_, connectorQueryCtx_));
-  VELOX_CHECK_NOT_NULL(
-      hiveWriterParameters,
-      "Hive data sink expects write parameters for Hive.");
-  writerParameters_.emplace_back(hiveWriterParameters);
-
-  auto writePath = fs::path(hiveWriterParameters->writeDirectory()) /
-      hiveWriterParameters->writeFileName();
+  auto writePath = fs::path(writerParameters->writeDirectory()) /
+      writerParameters->writeFileName();
   auto sink = dwio::common::DataSink::create(writePath);
   return std::make_unique<Writer>(
       options, std::move(sink), *connectorQueryCtx_->memoryPool());
 }
 
+void HiveDataSink::moveWriters() {
+  raw_vector<uint64_t> newIndices;
+
+  SelectivityVector validRows = SelectivityVector::empty(writers_.size());
+  for (vector_size_t i = 0; i < writers_.size(); i++) {
+    if (writers_[i]) {
+      validRows.setValid(i, true);
+    }
+  }
+  validRows.updateBounds();
+
+  // All valid values of partitionsVector_ are from previous addInput() that
+  // have been fed into partitionIdGenerator at least once, and thus assumed not
+  // to cause partitionIdGenerator to rehash.
+  partitionIdGenerator_.stableRun(partitionsVector_, validRows, newIndices);
+
+  uint64_t maxNewIndex = getMaxPartitionId(newIndices, validRows);
+
+  if (writers_.size() < maxNewIndex + 1) {
+    writers_.resize(maxNewIndex + 1);
+    writerParameters_.resize(maxNewIndex + 1);
+    partitionsVector_->resize(maxNewIndex + 1, false);
+  }
+  // Rehash monotonically makes the IDs of previous partition values larger. So
```

Comment:
This assumption tightly couples this code with VectorHasher's implementation. It would be nice to avoid that. Alternatively, we'd need to add unit tests that assert this behavior, so that if it ever changes, we'll get a clear signal.

Reply:
I chose to add unit tests to assert this behavior. W/o this assumption an alternative approach would be to have a second vector when moving the writers. Decided to trade away robustness for performance here.